File: viral-vtubers-backend/src/main/kotlin/com/viralvtubers/service/MailService.kt
package com.viralvtubers.service

import com.viralvtubers.graphql.data.ID
import com.viralvtubers.graphql.data.Mail
import com.viralvtubers.graphql.input.EditMailInput
import com.viralvtubers.graphql.input.SendMailInput

interface MailService {
    suspend fun getMail(userId: ID, mailId: ID): Mail

    suspend fun sendMail(userId: ID, input: SendMailInput): Mail

    suspend fun getInbox(userId: ID): List<Mail>

    suspend fun getSent(userId: ID): List<Mail>

    suspend fun editMail(userId: ID, input: EditMailInput): List<Mail>
}

File: viral-vtubers-frontend/src/app/services/mail.service.ts
import { Injectable } from '@angular/core';
import { map, Observable } from 'rxjs';
import { notificationQuery } from 'src/graphql/queries/self.query';
import {
  EditMailGQL,
  EditMailInput,
  InboxGQL,
  MailInboxFragmentFragment,
  MailOutboxFragmentFragment,
  OutboxGQL,
  SendMailGQL,
  SendMailInput,
} from 'src/schema/type';

@Injectable({
  providedIn: 'root',
})
export class MailService {
  inbox$?: Observable<MailInboxFragmentFragment[]>;
  outbox$?: Observable<MailOutboxFragmentFragment[]>;

  constructor(
    private inboxGQL: InboxGQL,
    private outboxGQL: OutboxGQL,
    private sendMailGQL: SendMailGQL,
    private editMailGQL: EditMailGQL
  ) {}

  getInbox() {
    this.inbox$ = this.inboxGQL
      .watch()
      .valueChanges.pipe(map((res) => res.data.self.inbox));
    return { query: this.inboxGQL, inbox$: this.inbox$ };
  }

  getOutbox() {
    this.outbox$ = this.outboxGQL
      .watch()
      .valueChanges.pipe(map((res) => res.data.self.sent));
    return { query: this.outboxGQL, outbox$: this.outbox$ };
  }

  sendMail(input: SendMailInput) {
    return this.sendMailGQL.mutate(
      { input },
      {
        refetchQueries: [{ query: notificationQuery }],
      }
    );
  }

  editMail(input: EditMailInput) {
    return this.editMailGQL.mutate(
      { input },
      {
        refetchQueries: [{ query: notificationQuery }],
      }
    );
  }
}
